astropy/modeling/utils.py

```diff
@@ -336,3 +341,67 @@ def add(self, operator_name, operator):
         self._set_value(key, operator)
 
         return key
+
+
+def copy_with_new_parameter_values(model, **parameters):
+    """
+    This function takes a model, as well as a dictionary of parameters, and
+    the final desired shape of the parameter arrays, and copies the model and
+    sets all the parameters to arrays with the desired shape. This is
+    essentially a workaround for the fact that if one has a scalar model, the
+    parameters cannot be changed to arrays. This issue is described in more
+    detail in https://github.com/astropy/astropy/issues/16593. Once that
+    issue is fixed, we should be able to remove this function and the
+    function below. Note that this preserves the constraints.
+    """
+    # Avoid circular imports
+    from astropy.modeling import CompoundModel, models
+
+    # TODO: detect case where new parameter shapes are not compatible with
+    # shape of existing parameters which will not get replaced.
+
+    shape = np.broadcast_shapes(*list(np.shape(value) for value in parameters.values()))
+
+    if isinstance(model, CompoundModel):
+        new_model = _compound_model_with_array_parameters(model, shape)
+        for name, value in parameters.items():
+            setattr(new_model, name, value)
+    else:
+        constraints = {}
+        for constraint in model.parameter_constraints:
+            constraints[constraint] = getattr(model, constraint)
+        # HACK: for some models, there are additional non-parameter arguments.
+        # We hard-code the fix for some models.
+        if isinstance(model, (models.Polynomial1D, models.Polynomial2D)):
+            args = (model.degree,)
```

Comment:
This is the major problem with this approach. We can't hard code these attributes for all models (I have custom DKIST ones for example), so we would need to generalise this or come up with a different solution.